### PR TITLE
added websocket & services

### PR DESCRIPTION
**Pull request by sharadregoti**

### CODE REVIEW

**Comment by YourTechBud** [comment]
Resolve the merge conflicts

**Comment by YourTechBud** [comment]
Where is the websocket file?

File: api/api.go [hunk 1]
```diff
@@ -58,8 +60,10 @@ func (api *API) Call(service, function string, params utils.M, timeout int) (*mo
 }
 
 // Service creates a Service instance
-func (api *API) Service(serviceName string) *service.Service {
-	return service.New(api.config, serviceName)
+func (api *API) Service(url string, config *config.Config, serviceName string) *service.Service {
+	val, _ := service.Init(config, serviceName, api.socket)
```
Comment (YourTechBud) [inline api/api.go]: Handle the error here. Let Service return an error as well

File: api/service/service.go [hunk 2]
```diff
@@ -1,126 +1,95 @@
...
-// 	id := uuid.NewV1().String()
-// 	var w transport.WebsocketConnection
-// 	var cb transport.CallBackFunctions
-// 	w.RegisterOnReconnectCallback(cb("service-register", {} ))
-
-// 	return &Service{config, service, id, make(map[string]ServiceFunction)}
-// }
+func Init(options *config.Config, serviceName string, client *websocket.Socket) (*Service, error) {
+	if err := client.ServiceRegister(serviceName); err != nil {
+		return nil, err
```
Comment (YourTechBud) [inline api/service/service.go]: Client should not have the service register method

File: api/service/service.go [hunk 3]
```diff
@@ -1,126 +1,95 @@
...
+	}
+	return &Service{
+		options: options,
+		service: serviceName,
+		client:  client,
+		function: functionInfo{
+			storeFunction: map[string]func(params interface{}, auth interface{}){},
```
Comment (YourTechBud) [inline api/service/service.go]: The function here should return a response as well

File: api/service/service.go [hunk 4]
```diff
@@ -1,126 +1,95 @@
...
+		},
+	}, nil
+	// service request ?
+}
 
-func (s *Service) RegisterFunc(funcName string, function ServiceFunction) {
-	s.funcs[funcName] = function
+func (s *Service) RegisterFunc(functionName string, function func(params interface{}, auth interface{}), response FunctionResponse) {
+	s.function.storeFunction[functionName] = function
+	s.function.response = response
```
Comment (YourTechBud) [inline api/service/service.go]: Whats the response for?

File: api/service/service.go [hunk 5]
```diff
@@ -1,126 +1,95 @@
...
-				}
+		s.mux.RLock()
+		funcInfo := s.function
+		s.mux.RUnlock()
+
+		function, ok := funcInfo.storeFunction[functionName]
+		if !ok {
+			s.client.WriteMessage <- websocket.WriteMessageStructure{
+				Type: websocket.ServiceRequest,
+				Data: websocket.FunctionsPayload{ID: req.ID, Error: "No function registered on the function"},
```
Comment (YourTechBud) [inline api/service/service.go]:
```suggestion
				Data: websocket.FunctionsPayload{ID: req.ID, Error: "No function registered on the service"},
```

File: api/service/service.go [hunk 6]
```diff
@@ -1,126 +1,95 @@
...
 			}
-			close(c)
-		} else {
-			log.Println("Not connected. Attempting to Reconnect...")
+		}
 
+		function(params, auth)
```
Comment (YourTechBud) [inline api/service/service.go]: The function should return a response and an error

File: api/service/service.go [hunk 7]
```diff
@@ -1,126 +1,95 @@
...
+
+		switch funcInfo.response.Type {
+		case "response":
+			s.mux.Lock()
+			s.client.WriteMessage <- websocket.WriteMessageStructure{
+				Type: websocket.ServiceRequest,
+				Data: websocket.FunctionsPayload{ID: req.ID, Params: req.Params},
```
Comment (YourTechBud) [inline api/service/service.go]: Params here will be the response returned by the function

File: api/transport/websocket/structures.go [hunk 8]
```diff
@@ -0,0 +1,33 @@
+package websocket
+
+type websocketOptions struct {
+	projectId string
+	token     string
+}
+
+type WriteMessageStructure struct {
```
Comment (YourTechBud) [inline api/transport/websocket/structures.go]: You do not need this. Just use the Message

File: api/transport/websocket/structures.go [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+	Type string
+	Data interface{}
+}
+
+type ServiceRegisterRequest struct {
```
Comment (YourTechBud) [inline api/transport/websocket/structures.go]: Put all stucts in the model folder

File: api/transport/websocket/websocket.go [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+
+const ServiceRegister string = "service-register"
+const ServiceRequest string = "service-request"
+
+type Socket struct {
+	url             string
+	options         websocketOptions
+	isConnected     bool
+	channels        map[string]chan FunctionsPayload
+	serviceRegister map[string]chan map[string]bool
```
Comment (YourTechBud) [inline api/transport/websocket/websocket.go]: This shouldn't be here

File: api/transport/websocket/websocket.go [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+	s.isConnected = value
+	s.mux.Unlock()
+}
+
+func (s *Socket) RegisterChannel(name string, channel chan FunctionsPayload) {
+	_, ok := s.channels[name]
+	if !ok {
+		s.channels[name] = channel
+	} else {
+		// error  already exists
```
Comment (YourTechBud) [inline api/transport/websocket/websocket.go]: Return an error here

**Review by YourTechBud — CHANGES_REQUESTED** [review]
Im not reviewing the websocket.go file since its api isn't similar to the one we had discussed. You need to have the same functions as there was in the js library

File: api/service/service.go [hunk 12]
```diff
@@ -1,126 +1,111 @@
...
-	Params   []byte `json:"params"`
-	Function string `json:"function"`
-	Error    string `json:"error"`
+type function func(params interface{}, auth map[string]interface{}) (interface{}, error)
+
+func Init(options *config.Config, serviceName string, client *websocket.	Socket) (*Service, error) {
+	client.RegisterOnReconnectCallback(func() {
+		v, err := client.Request(websocket.ServiceRegister, model.ServiceRegisterRequest{Service: serviceName, Project: options.Project, Token: options.Token})
+		if err != nil {
+			log.Println(err)
```
Comment (YourTechBud) [inline api/service/service.go]:
```suggestion
			log.Println("Service register error:", err)
```

File: api/service/service.go [hunk 13]
```diff
@@ -1,126 +1,111 @@
...
+			return
+		}
+		responseData, ok := v.(map[string]bool)
```
Comment (YourTechBud) [inline api/service/service.go]: `v` will be a `map[string]interface{}`

File: api/service/service.go [hunk 14]
```diff
@@ -1,126 +1,111 @@
...
+		if !ok {
+			log.Println("service initialization wrong type found")
+			return
+		}
+		data, ok := responseData["ack"]
+		if ok {
+			log.Println("service initialization didn't received acknowledgement from knowledge")
```
Comment (YourTechBud) [inline api/service/service.go]:
```suggestion
			log.Println("Service register error: Didn't received acknowledgement from server")
```

File: api/service/service.go [hunk 15]
```diff
@@ -1,126 +1,111 @@
...
+		if err != nil {
+			log.Println(err)
+			return
+		}
+		responseData, ok := v.(map[string]bool)
+		if !ok {
+			log.Println("service initialization wrong type found")
+			return
+		}
+		data, ok := responseData["ack"]
```
Comment (YourTechBud) [inline api/service/service.go]:
```suggestion
		ack, ok := responseData["ack"]
```

File: api/service/service.go [hunk 16]
```diff
@@ -1,126 +1,111 @@
...
-	Type     string `json:"type"`
-	Id       string `json:"id"`
-	Service  string `json:"service"`
-	Params   []byte `json:"params"`
-	Function string `json:"function"`
-	Error    string `json:"error"`
+type function func(params interface{}, auth map[string]interface{}) (interface{}, error)
+
+func Init(options *config.Config, serviceName string, client *websocket.	Socket) (*Service, error) {
+	client.RegisterOnReconnectCallback(func() {
```
Comment (YourTechBud) [inline api/service/service.go]: Register this callback in Start()

File: api/service/service.go [hunk 17]
```diff
@@ -1,126 +1,111 @@
...
+		log.Println("Service started successfully")
+	})
+
+	return &Service{options: options, service: serviceName, client: client, function: map[string]function{}}, nil
 }
 
-func New(config *config.Config, service string) *Service {
-	return nil
+func (s *Service) RegisterFunc(functionName string, fn function) {
+	s.function[functionName] = fn
```
Comment (YourTechBud) [inline api/service/service.go]: Wrap this within a lock

File: api/service/service.go [hunk 18]
```diff
@@ -1,126 +1,111 @@
...
+	reqMap, ok := data.(map[string]interface{})
+	if !ok {
+		log.Println("service request type not found")
+	}
 
-// 	return &Service{config, service, id, make(map[string]ServiceFunction)}
-// }
+	req := model.FunctionsPayload{}
+	if err := mapstructure.Decode(reqMap, &req); err != nil {
+		log.Println(err)
```
Comment (YourTechBud) [inline api/service/service.go]: you should print the error with an appropriate caption. Also, you forgot to send the error and then return.

File: api/service/service.go [hunk 19]
```diff
@@ -1,126 +1,111 @@
...
-	// function := req.Function
-	// params := req.Params
-	// auth := req.Auth
-	// if (!auth || Object.keys(auth) == 0) auth = null;
+	if auth == nil || len(auth) == 0 {
+		auth = nil
+	}
+
+	s.mux.RLock()
+	funcInfo := s.function
```
Comment (YourTechBud) [inline api/service/service.go]: make a function which returns the function and a bool. Wrap that function within a lock. Put it in `helpers.go`

File: api/service/service.go [hunk 20]
```diff
@@ -1,126 +1,111 @@
...
+	s.mux.RUnlock()
 
+	function, ok := funcInfo[functionName]
+	if !ok {
+		s.client.Send(websocket.ServiceRequest, model.FunctionsPayload{ID: req.ID, Error: "No function registered on the service"})
+		return
+	}
+	response, err := function(params, auth)
+	if err != nil {
+		log.Println(err)
```
Comment (YourTechBud) [inline api/service/service.go]: Log error with the caption. Send the error and then return

File: api/service/service.go [hunk 21]
```diff
@@ -1,126 +1,111 @@
...
-					}
-				}
-			}
-			close(c)
-		} else {
-			log.Println("Not connected. Attempting to Reconnect...")
+func (s *Service) Start() error {
+	s.client.RegisterCallback(websocket.ServiceRequest, s.serviceRequest)
+
+	v, err := s.client.Request(websocket.ServiceRegister, model.ServiceRegisterRequest{Service: s.service, Project: s.options.Project, Token: s.options.Token,})
```
Comment (YourTechBud) [inline api/service/service.go]: make a function names serviceRegister in helpers to take care of this. Code isn't DRY right now.

File: api/transport/websocket/helpers.go [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+
+import (
+	"github.com/gorilla/websocket"
+	"github.com/spaceuptech/space-api-go/api/model"
+)
+
+func (s *Socket) setSocket(socket *websocket.Conn) {
+	s.mux.Lock()
+	s.Socket = socket
+	s.mux.Unlock()
```
Comment (YourTechBud) [inline api/transport/websocket/helpers.go]: exec all unlocks in defer

File: api/transport/websocket/helpers.go [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+func (s *Socket) setWriterChannel(ch chan model.WebsocketMessage) {
+	s.mux.Lock()
+	s.SendMessage = ch
+	s.mux.Unlock()
+}
+
+func (s *Socket) RegisterCallback(Type string, function func(data interface{})) {
```
Comment (YourTechBud) [inline api/transport/websocket/helpers.go]: All exported functions must have a comment and be in a file named `operations.go`

File: api/transport/websocket/websocket.go [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+	token     string
+}
+
+type Socket struct {
+	url                  string
+	isConnected          bool
+	connectedOnce        bool
+	options              websocketOptions
+	pendingMsg           []model.WebsocketMessage
+	Socket               *websocket.Conn
```
Comment (YourTechBud) [inline api/transport/websocket/websocket.go]: no need to export any member variable

File: api/transport/websocket/websocket.go [hunk 25]
```diff
@@ -0,0 +1,162 @@
...
+	conn, _, err := websocket.DefaultDialer.Dial(s.url, nil)
+	if err != nil {
+		log.Println(err)
+		return err
+	}
+
+	s.setSocket(conn)
+	s.setConnected(true)
+
+	if s.getConnectedOnce() {
```
Comment (YourTechBud) [inline api/transport/websocket/websocket.go]: rename this to `isConnectedOnce`

File: api/transport/websocket/websocket.go [hunk 26]
```diff
@@ -0,0 +1,162 @@
...
+		for _,value := range s.onReconnectCallbacks {
+			value()
+		}
+	}
+
```
Comment (YourTechBud) [inline api/transport/websocket/websocket.go:70]: Set isConnected to true

File: api/transport/websocket/websocket.go [hunk 27]
```diff
@@ -0,0 +1,162 @@
...
+	if len(s.pendingMsg) != 0 {
+		for _, payload := range s.pendingMsg {
+			s.Socket.WriteJSON(payload)
+		}
+		s.pendingMsg = []model.WebsocketMessage{}
+	}
+
+	writeMessage := make(chan model.WebsocketMessage)
```
Comment (YourTechBud) [inline api/transport/websocket/websocket.go]: Make this inside the Init method. You dont need to close this channel at all

File: api/transport/websocket/websocket.go [hunk 28]
```diff
@@ -0,0 +1,162 @@
...
+	defer close(writeMessage)
+	s.setWriterChannel(writeMessage)
+
+	// create a websocket reader & writer
+	go s.writerRoutine()
```
Comment (YourTechBud) [inline api/transport/websocket/websocket.go]: You need a single writer routine for the entire socket object

File: api/transport/websocket/websocket.go [hunk 29]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+func (s *Socket) read() {
+	msg := model.WebsocketMessage{}
+	for {
+		if s.getConnected() {
+			if err := s.Socket.ReadJSON(&msg); err != nil {
+				log.Println(err)
+				s.setConnected(false)
+				continue
```
Comment (YourTechBud) [inline api/transport/websocket/websocket.go:142]: sleep for 5 seconds here

File: api/transport/websocket/websocket.go [hunk 30]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+func (s *Socket) Send(Type string, data interface{}) string {
+	id := uuid.NewV1().String()
+	s.SendMessage <- model.WebsocketMessage{ID: id, Type: Type, Data: data}
+	return id
+}
+
+func (s *Socket) read() {
+	msg := model.WebsocketMessage{}
```
Comment (YourTechBud) [inline api/transport/websocket/websocket.go]: create this msg object inside the loop

File: api/transport/websocket/websocket.go [hunk 31]
```diff
@@ -0,0 +1,162 @@
...
+			}
+		} else {
+			if err := s.connect(); err != nil {
+				log.Println(err)
+				continue
+			}
+		}
+
+		if msg.ID != "" {
+			cb, ok := s.registerCallbackMap[msg.ID]
```
Comment (YourTechBud) [inline api/transport/websocket/websocket.go]: have a helper function to read the cb. You need a read lock before doing this operation

File: api/transport/websocket/websocket.go [hunk 32]
```diff
@@ -0,0 +1,162 @@
...
+			if ok {
+				go cb(msg.Data)
+				s.unregisterCallback(msg.ID)
+			}
+		}
+
+		cb, ok := s.registerCallbackMap[msg.Type]
```
Comment (YourTechBud) [inline api/transport/websocket/websocket.go]: Read lock needed here

**Review by YourTechBud — CHANGES_REQUESTED** [review]

File: api/service/service.go (hunk 12 above)
Comment (sharadregoti) [reply api/service/service.go]: done

File: api/service/service.go (hunk 13 above)
Comment (sharadregoti) [reply api/service/service.go]: done

File: api/service/service.go (hunk 14 above)
Comment (sharadregoti) [reply api/service/service.go]: done

File: api/service/service.go (hunk 15 above)
Comment (sharadregoti) [reply api/service/service.go]: done

File: api/service/service.go (hunk 16 above)
Comment (sharadregoti) [reply api/service/service.go]: done

File: api/service/service.go (hunk 17 above)
Comment (sharadregoti) [reply api/service/service.go]: done

File: api/service/service.go (hunk 18 above)
Comment (sharadregoti) [reply api/service/service.go]: done

File: api/service/service.go (hunk 19 above)
Comment (sharadregoti) [reply api/service/service.go]: done

File: api/service/service.go (hunk 20 above)
Comment (sharadregoti) [reply api/service/service.go]: done

File: api/service/service.go (hunk 21 above)
Comment (sharadregoti) [reply api/service/service.go]: done

File: api/transport/websocket/helpers.go (hunk 22 above)
Comment (sharadregoti) [reply api/transport/websocket/helpers.go]: done

File: api/transport/websocket/helpers.go (hunk 23 above)
Comment (sharadregoti) [reply api/transport/websocket/helpers.go]: done

File: api/transport/websocket/websocket.go (hunk 24 above)
Comment (sharadregoti) [reply api/transport/websocket/websocket.go]: done

File: api/transport/websocket/websocket.go (hunk 25 above)
Comment (sharadregoti) [reply api/transport/websocket/websocket.go]: done

File: api/transport/websocket/websocket.go (hunk 26 above)
Comment (sharadregoti) [reply api/transport/websocket/websocket.go:70]: done

File: api/transport/websocket/websocket.go (hunk 27 above)
Comment (sharadregoti) [reply api/transport/websocket/websocket.go]: done

File: api/transport/websocket/websocket.go (hunk 28 above)
Comment (sharadregoti) [reply api/transport/websocket/websocket.go]: done

File: api/transport/websocket/websocket.go (hunk 29 above)
Comment (sharadregoti) [reply api/transport/websocket/websocket.go:142]: done

File: api/transport/websocket/websocket.go (hunk 30 above)
Comment (sharadregoti) [reply api/transport/websocket/websocket.go]: done

**Review by YourTechBud — APPROVED** [review]

**Pull request merged**